File: apps/rust/test_app/src/primes.rs
// Copyright (C) 2023 Javier Albiero (jalbiero)
// Distributed under the MIT License (see the accompanying LICENSE file
// or go to http://opensource.org/licenses/MIT).


// Inline attributes are just to be on pair with C++ implementation (see primes.hpp)

#[inline]
fn is_prime(n: i32) -> bool {
    if n <= 1 {
        return false;
    }

    if n == 2 {
        return true;
    }

    // the only even prime is 2
    if n % 2 == 0 {
        return false;
    }

    // Trial division (https://en.wikipedia.org/wiki/Prime_number#Trial_division)
    let limit = (n as f32).sqrt() as i32 + 1;

    for i in 3..limit {
        if n % i == 0 {
            return false;
        }
    }

    true
}

#[inline]
pub fn get_primes_less_than(limit: i32) -> Vec<i32> {
    let mut result: Vec<i32> = Vec::new();

    for i in 2..limit {
        if is_prime(i) {
            result.push(i);
        }
    }

    result
}

#[inline]
pub fn count_primes_less_than(limit: i32) -> i32 {
    let mut result: i32 = 0;

    for i in 2..limit {
        if is_prime(i) {
            result += 1;
        }
    }

    result
}

File: apps/csharp/test_app/src/Primes.cs
internal static class Primes {
    public static bool IsPrime(int num) {
        if (num <= 1) return false;
        if (num == 2) return true;
        if (num % 2 == 0) return false; // the only even prime is 2

        // Trial division (https://en.wikipedia.org/wiki/Prime_number#Trial_division)
        var limit = Math.Sqrt(num);

        for (int i=3; i<=limit; i++) {
            if (num % i == 0) return false;
        }

        return true;
    }

    public static IList<int> GetPrimesLessThan(int limit) {
        var result = new List<int>();

        for (int i=2; i<limit; i++) {
            if (IsPrime(i)) result.Add(i);
        }

        return result;
    }

    public static int CountPrimesLessThan(int limit) {
        int result = 0;

        for (int i=2; i<limit; i++) {
            if (IsPrime(i)) result++;
        }

        return result;
    }
}
